File: 프로그래머스/LV1/코딩테스트 연습/연습문제/명예의 전당 (1)/명예의 전당 (1).cpp
#include <vector>
#include <queue>
using namespace std;
typedef priority_queue<int, vector<int>, greater<int>> HEAP;
vector<int> solution(int k, vector<int> score) 
{
    vector<int> answer, hall_of_fame;
    HEAP heap(hall_of_fame.begin(), hall_of_fame.end());

    for(int s : score)
    {
        heap.push(s);
        if(heap.size() > k)heap.pop();
        answer.push_back(heap.top());
    }
    return answer;
}

File: 프로그래머스/LV1/코딩테스트 연습/연습문제/명예의 전당 (1)/명예의 전당 (1).py
from heapq import heapify, heappush, heappop

def solution(k, score):
    answer = []
    hall_of_fame = []
    heapify(hall_of_fame)

    for s in score:
        heappush(hall_of_fame, s)
        if len(hall_of_fame) > k:
            heappop(hall_of_fame)
            
        answer.append(hall_of_fame[0])
    return answer
